### ISSUE TITLE
NServiceBus OutOfMemoryException Using Azure Service Bus Transport & SQL Server for Outbox Feature

### ISSUE THREAD

**Issue by trupnick** [Particular/NServiceBus#6598]
I'm looking for some pointers to figure out root cause of an out of memory exception we are seeing with our NServiceBus setup. We're using Azure Service Bus for the transport, and an Azure SQL database for the outbox feature. We run the app that hosts NServiceBus in an Azure App Service on .NET Core 6.0 with a 32-bit processor architecture. 

While we were getting the out of memory exceptions the memory for the app service seemed to be doing fine, the Azure SQL database used for the outbox feature was performing well during the out of memory exceptions too. We use NServiceBus 7.3.  

[StackOverflow Post](https://stackoverflow.com/questions/74521234/how-can-i-troubleshoot-nservicebus-outofmemoryexception-using-azure-service-bus)

[Azure App Service Performance Counters](https://i.stack.imgur.com/V0f11.png)

> StackTrace:
>
> System.OutOfMemoryException:
>
> at System.Threading.Thread.StartInternal (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Thread.StartCore (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Thread.Start (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Thread.Start (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection (Microsoft.Data.SqlClient, Version=4.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection (Microsoft.Data.SqlClient, Version=4.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal (Microsoft.Data.SqlClient, Version=4.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionClosed.TryOpenConnection (Microsoft.Data.SqlClient, Version=4.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlConnection.TryOpen (Microsoft.Data.SqlClient, Version=4.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlConnection.InternalOpenAsync (Microsoft.Data.SqlClient, Version=4.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Extensions+<OpenConnection>d__4.MoveNext (NServiceBus.Persistence.Sql, Version=6.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c: /_/src/SqlPersistence/Extensions.cs:46)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at OutboxPersister+<Get>d__9.MoveNext (NServiceBus.Persistence.Sql, Version=6.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c: /_/src/SqlPersistence/Outbox/OutboxPersister.cs:75)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at NServiceBus.TransportReceiveToPhysicalMessageConnector+<Invoke>d__1.MoveNext (NServiceBus.Core, Version=7.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at NServiceBus.MainPipelineExecutor+<Invoke>d__1.MoveNext (NServiceBus.Core, Version=7.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at NServiceBus.Transport.AzureServiceBus.MessagePump+<InnerProcessMessage>d__23.MoveNext (NServiceBus.Transport.AzureServiceBus, Version=1.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c)

NServiceBus endpoint config:

```
public static IHostBuilder ConfigureNServiceBus(this IHostBuilder hostBuilder, string connectionString)
{
    hostBuilder.UseNServiceBus(builderContext =>
    {
        var endpointConfiguration = new EndpointConfiguration("MyEndpoint");
        endpointConfiguration.AddDeserializer<NewtonsoftSerializer>();
        endpointConfiguration.UseSerialization<NewtonsoftSerializer>();
        endpointConfiguration.DisableFeature<AutoSubscribe>();
        endpointConfiguration.SendFailedMessagesTo("error");
        endpointConfiguration.AuditProcessedMessagesTo("audit");
        endpointConfiguration.DisableFeature<AutoSubscribe>();

        endpointConfiguration.UseTransport<AzureServiceBusTransport>()
            .ConnectionString(builderContext.Configuration.GetAzureServiceBusConnectionString());

        var persistence = endpointConfiguration.UsePersistence<SqlPersistence>();

        persistence.SqlDialect<SqlDialect.MsSqlServer>().Schema(Constants.NServiceBusSchema);

        persistence.ConnectionBuilder(() => new SqlConnection(connectionString));

        endpointConfiguration.EnableOutbox();

        return endpointConfiguration;
    });

    return hostBuilder;
}
```

[NServiceBus Nugets](https://i.stack.imgur.com/IG61z.png)

We switched the app to run with a 64-bit processor architecture and this resolved the issue. We have several other apps that use NServiceBus with the same transport and outbox feature and they run fine with a 32-bit processor architecture.  

Does NServiceBus have a set of minimum hardware requirements? I didn't find these in the [Particular Docs](https://docs.particular.net/nservicebus/upgrades/supported-platforms)

**Comment by lailabougria** [comment]
Hey @trupnik, you should be able to run just fine on 32-bit and 64-bit in NServiceBus terms.  Memory issues could be caused by many different reasons. 

I saw that you're running on an unsupported version of NServiceBus, and most of the downstream packages you're using are also outdated. Have you validated whether the same behavior occurs on the [latest version](https://docs.particular.net/nservicebus/upgrades/supported-versions)?

Can you share some more information about the scenario? Do you see this occur after the endpoint has been running for some time? Or does it occur almost immediately after startup? Which specs are you currently working with, or rather, which service plan are you currently using?

Can you try to use some memory profiler like [dotMemory](https://www.jetbrains.com/dotmemory/) to pinpoint any instances of high memory allocation and memory that's been kept around for a long time? If you don't have the tooling available or can't get a trial version, sharing a memory dump would also be helpful.

Switching to 64-bit will work for now as the [process can now allocate more memory](https://learn.microsoft.com/en-us/dotnet/api/system.outofmemoryexception?view=net-7.0), but you might run into issues again in the future, so it's important to chase down the root cause.

PS: If you can't share any of the above here on the issue, feel free to send this information to us through [support](https://particular.net/support).

**Comment by trupnick** [comment]
Hi @lailabougria - thanks for the response!

Good call on updating the version.  I didn't notice that we were many version behind the supported version.  I'll get that updated.

The application was running without error for many months on our production app service plan which is an [S3:10](https://azure.microsoft.com/en-us/pricing/details/app-service/windows/#pricing).  The app service plan is shared by multiple app services.  There were some other apps added to the plan recently that with our scaling rules scaled from 7 instances to 10.  My team deployed a small code change to our service which was the point at which we started seeing the out of memory exceptions.  We rolled-back the the previous version of the app but the out of memory exception persisted.  We then switched the app to 64-bit and the exceptions stopped.

100% agree.  I definitely need to get to the root cause because switching to 64-bit is just a band aid.  Over the next couple of days I'm going to work to do memory profiling to see if there any large objects allocated in our app.  I will report back with my findings.
